### ISSUE TITLE
Silent feature name loss in xgboost 3.1.3 + pandas 3.0: lazy_isinstance fails to recognize DataFrames

### ISSUE THREAD

**Issue by JMerrill-Fairness** [dmlc/xgboost#12000]
We hit a subtle but catastrophic manifestation of this issue worth documenting: xgboost 3.1.3 silently drops feature names when training with pandas 3.0 DataFrames. booster.feature_names returns None after training, with no error or warning.

Root cause: _is_pandas_df in compat.py checks cls.__module__ == "pandas.core.frame", but pandas 3.0 changed DataFrame.__module__ to "pandas". The DataFrame falls through dispatch_data_backend to _has_array_protocol / np.asarray(), silently stripping all column metadata.

Minimal repro:

import pandas as pd
import xgboost as xgb

print(f"xgboost={xgb.__version__}, pandas={pd.__version__}")

df = pd.DataFrame({"feat_a": [0, 1, 0, 1], "feat_b": [1, 0, 1, 0]})
y = [0, 1, 0, 1]

# Via sklearn API
model = xgb.XGBClassifier(n_estimators=2, max_depth=1)
model.fit(df, y)
print("booster.feature_names:", model.get_booster().feature_names)
# pandas 2.x: ['feat_a', 'feat_b']
# pandas 3.0: None  <-- silent data loss

# Via DMatrix API
dmat = xgb.DMatrix(df, label=y)
print("dmatrix.feature_names:", dmat.feature_names)
# pandas 2.x: ['feat_a', 'feat_b']
# pandas 3.0: ['f0', 'f1']  <-- generic names, no error

Versions: xgboost 3.1.3, pandas 3.0.0, Python 3.11.14

Workaround: Pass feature_names explicitly to DMatrix(), or set model.get_booster().feature_names after .fit():

dmat = xgb.DMatrix(df, label=y, feature_names=[str(c) for c in df.columns])

# or after sklearn .fit():
model.fit(df, y)
model.get_booster().feature_names = [str(c) for c in df.columns]

**Comment by trivialfis** [comment]
Thank you for sharing, we are preparing a new release to support pandas 3.0

**Comment by trivialfis** [comment]
3.2 is available now, ran the test and got:
```
xgboost=3.2.0, pandas=3.0.0
booster.feature_names: ['feat_a', 'feat_b']
dmatrix.feature_names: ['feat_a', 'feat_b']
```
